File: typescript/tests/webhooks/verify.test.ts
import { describe, it, expect } from "vitest";
import { verifyWebhookSignature, signWebhookPayload } from "../../src/webhooks/verify.js";

describe("signWebhookPayload", () => {
  it("produces deterministic signatures", () => {
    const sig1 = signWebhookPayload("test", "secret");
    const sig2 = signWebhookPayload("test", "secret");
    expect(sig1).toBe(sig2);
  });

  it("produces different signatures for different payloads", () => {
    const sig1 = signWebhookPayload("payload1", "secret");
    const sig2 = signWebhookPayload("payload2", "secret");
    expect(sig1).not.toBe(sig2);
  });

  it("produces different signatures for different secrets", () => {
    const sig1 = signWebhookPayload("test", "secret1");
    const sig2 = signWebhookPayload("test", "secret2");
    expect(sig1).not.toBe(sig2);
  });
});

describe("verifyWebhookSignature", () => {
  it("round-trips with signWebhookPayload", () => {
    const payload = '{"id":1,"kind":"todo_created"}';
    const secret = "test-secret-key";
    const signature = signWebhookPayload(payload, secret);

    expect(verifyWebhookSignature(payload, signature, secret)).toBe(true);
  });

  it("works with Buffer payloads", () => {
    const payload = Buffer.from('{"id":1}');
    const secret = "key";
    const signature = signWebhookPayload(payload, secret);

    expect(verifyWebhookSignature(payload, signature, secret)).toBe(true);
  });

  it("rejects invalid signatures", () => {
    expect(verifyWebhookSignature("data", "wrong", "secret")).toBe(false);
  });

  it("returns false for empty secret", () => {
    expect(verifyWebhookSignature("data", "sig", "")).toBe(false);
  });

  it("returns false for empty signature", () => {
    expect(verifyWebhookSignature("data", "", "secret")).toBe(false);
  });

  it("returns false when signature length differs", () => {
    expect(verifyWebhookSignature("data", "short", "secret")).toBe(false);
  });
});

File: ruby/test/basecamp/webhooks/verify_test.rb
# frozen_string_literal: true

require "test_helper"

class Basecamp::Webhooks::VerifyTest < Minitest::Test
  def test_round_trip
    payload = '{"id":1,"kind":"todo_created"}'
    secret = "test-secret-key"

    signature = Basecamp::Webhooks::Verify.compute_signature(payload: payload, secret: secret)
    assert Basecamp::Webhooks::Verify.valid?(payload: payload, signature: signature, secret: secret)
  end

  def test_rejects_wrong_signature
    payload = '{"id":1}'
    secret = "test-secret"

    assert_not Basecamp::Webhooks::Verify.valid?(payload: payload, signature: "wrong", secret: secret)
  end

  def test_rejects_empty_secret
    assert_not Basecamp::Webhooks::Verify.valid?(payload: "data", signature: "sig", secret: "")
    assert_not Basecamp::Webhooks::Verify.valid?(payload: "data", signature: "sig", secret: nil)
  end

  def test_rejects_empty_signature
    assert_not Basecamp::Webhooks::Verify.valid?(payload: "data", signature: "", secret: "secret")
    assert_not Basecamp::Webhooks::Verify.valid?(payload: "data", signature: nil, secret: "secret")
  end

  def test_compute_signature_deterministic
    payload = "test"
    secret = "key"
    sig1 = Basecamp::Webhooks::Verify.compute_signature(payload: payload, secret: secret)
    sig2 = Basecamp::Webhooks::Verify.compute_signature(payload: payload, secret: secret)
    assert_equal sig1, sig2
  end
end
